trieste/space.py

```diff
@@ -331,3 +331,149 @@ def __mul__(self, other: Box) -> Box:
 
     def __deepcopy__(self, memo: dict[int, object]) -> Box:
         return self
+
+
+class OrdinalSearchSpace(Box):
+    r"""
+    Ordinal search space representing a discrete :math:`D`-dimensional box in
+    :math:`\mathbb{R}^D`. This class inherit :class:`Box` and generates evenly-spaced values
+    in an interval for each dimension. For example randomly sample from this search space
+    will give:
+
+        >>> tf.random.set_seed(1945)
+        >>> ordinalsp = OrdinalSearchSpace([-2.0, -2.0], [2.0, 2.0], [0.2, 0.3])
+        >>> sample = ordinalsp.sample(5)
+        >>> sample.numpy()
+        array([[ 0. , -0.6],
+               [ 0.6, -0.3],
+               [-1.2,  1.2],
+               [-0.4,  1.5],
+               [ 1.2,  0. ]])
+
+    """
+
+    def __init__(
+        self,
+        lower: Sequence[float] | TensorType,
+        upper: Sequence[float] | TensorType,
+        stepsizes: Sequence[float] | TensorType,
+    ):
+
+        r"""
+        If ``lower``, ``upper`` and ``stepsizes`` are `Sequence`\ s of floats
+        (such as lists or tuples), they will be converted to tensors of dtype `tf.float64`.
+        if ``lower`` and ``upper`` are not multiple of ``stepsizes`` it will be replaced by
+        smallest and largest multiples inside previous ``lower`` and ``upper``.
+
+        :param lower: The lower (inclusive) bounds of the Ordinal Space. Must have shape [D]
+            for positive D, and if a tensor, must have float type.
+        :param upper: The upper (inclusive) bounds of the Ordinal Space. Must have shape [D]
+            for positive D, and if a tensor, must have float type.
+        :param stepsizes: The step size of each dimension of the Ordinal Space. Must have
+            shape [D] for positive D, and if a tensor, must have float type.
+        :raise ValueError (or InvalidArgumentError): If any of the following are true:
+
+            - ``lower``, ``upper`` and ``stepsizes`` have invalid shapes.
+            - ``lower``, ``upper`` and ``stepsizes`` do not have the same floating point type.
+            - ``upper`` is not greater than ``lower`` across all dimensions.
+        """
+
+        super().__init__(lower, upper)
```

Comment:
This doesn't make sense here if we're potentially changing the lower/upper bounds later on. Maybe move it to the end of the function and pass in the updated values of lower/upper instead? (you can use normal local variables for those too, as it's `super()`'s responsibility to set the private variables)

This will also catch the case where lower is rounded up and upper is rounded down, leading to lower > upper. You should probably test this by checking that e.g. the following raises an error:

```python
OrdinalSearchSpace([0.1], [0.9], [1.0])
```

Reply:
tried to do that, just realized that `super()` also responsible to change the dtype of _upper and _lower, thus if I put it at the end of the __init__ the dtype equality tests fails. Should we add Box's upper and lower dtype conversion here? or we can re do lower>upper assert here since it less redundant. what do you think? 
I am not quite sure how to make sure that we always have same dtype.

Reply:
Another possibility then is to leave super() where it is, but then call it _again_ at the end of the function with the rounded bounds?

Reply:
sorry wasn't aware on your last comment. I'll try it.

Reply:
Done!